### PR TITLE
Add autoscale check for GPU

### LINKED ISSUE #407: [Feature Request]: GPUs for Auto Scaling
### Feature description

We need to support a GPU dropdown that detects availability of auto-scaling and present that option to the user.

Jeff noted some of his thoughts on how the GPU dropdown will work in #301 ([comment](https://github.com/opendatahub-io/odh-dashboard/issues/301#issuecomment-1192576547)).

When the user cannot get the GPU count or in fact there are zero usage on the given server, we need to show the disabled state still. Otherwise we'll need to allow them to scale up to whatever the max GPU allowed is (but making it clear that some are available - if any - and others are unavailable and doing so will cause a wait for availability).

### Describe alternatives you've considered

We may need UX to look at what the items in the dropdown say -- maybe something like:
```
0
1
2 (unavailable - can auto-scale)
```
May be a bit winded, perhaps a dropdown or something also.

### Anything else?

RHODS ticket: https://issues.redhat.com/browse/RHODS-4617

### PR DESCRIPTION
**Pull request by maroroman**
Resolves #407 

## Description
This PR adds logic to check the MachineAutoscaler CRs for the maximum amount of scalable GPU nodes.

## How Has This Been Tested?

Testing was done on a OSD cluster with a gpu node with autoscaling enabled.

1. Use following dashboard image: quay.io/mroman_redhat/odh-dashboard:dev-gpu-3
2. The dashboard detects all MachineAutoscaler CRs and loops through their MachineSets which contain the `machine.openshift.io/GPU` value which should contain the true gpu amount (correct me if my assumption is wrong)

Resulting backend response was this: 
image

The value is used by the frontend to extend the dropdown if the scalable gpus are over the amount of gpus reported by prometheus metrics.

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by maroroman** [comment]
@andrewballantyne Should I add the autoscaler value to the available gpus and let the dropdown just show a higher number without a UX difference?

**Comment by andrewballantyne** [comment]
@maroroman see [Jeff's comment here](https://github.com/opendatahub-io/odh-dashboard/issues/301#issuecomment-1192576547), I think the answer to your question is this sentence:
> The number in the drop down should be the max available on any single node

No UX difference, just show the numbers.

**Comment by maroroman** [comment]
Sadly the ClusterAutoscaler i not used, however I will rewrite the branch to use MachineSets and MachineAutoscalers. I will also modify the gpuInfo object to contain information about each autoscaler to accurately display the amount of gpus (i.e. the user can scale 1 2-gpu node, 2 1-gpu nodes and one 2-gpu node is already reported by prometheus, so the value should be max 2 in the dropdown and it should let 2 users choose up to 2 and 2 users up to 1)

File: frontend/src/pages/notebookController/screens/server/GPUSelectField.tsx
Comment (andrewballantyne) [inline frontend/src/pages/notebookController/screens/server/GPUSelectField.tsx:64]: Why not make this part of the fetch logic? This hook only runs once per fetch call, no?

**Review by andrewballantyne — APPROVED** [review]
Couple nits, nothing too crazy.

/hold

I'm going to approve this as you DM'ed me it is working and I don't have a cluster to test this. 

Maybe @cfchase can test it out for a double verification.

**Comment by andrewballantyne** [comment]
Issue was found... un-lgtm-ing

**Comment by andrewballantyne** [comment]
Tested on Maros' cluster, looks to be working as intended.

File: backend/src/routes/api/gpu/gpuUtils.ts
Comment (cfchase) [inline backend/src/routes/api/gpu/gpuUtils.ts:16]: If the only gpus are on a machine pool that is scaled to 0, you're not going to have any pods, the daemonset `nvidia-dcgm-exporter` would have scaled them to 0.  If you're looking for the daemonset you should query for that.  You could alternatively query for the subscription `nvidia-gpu-operator` and check its `csv` the same way the tiles do.  You probably need to look into the recommended way to do this.

Comment (andrewballantyne) [reply backend/src/routes/api/gpu/gpuUtils.ts:16]: FWIW, he does a check if he found scaling information he doesn't need the pod information. [code link](https://github.com/opendatahub-io/odh-dashboard/pull/573/files#diff-fe048868b84dce849bd698d4f237f445ecc70661f9ca1842d9a9429b4c2c37feR50)

Comment (cfchase) [reply backend/src/routes/api/gpu/gpuUtils.ts:16]: sorry, was on looking at an older version before your last commit

File: backend/src/routes/api/gpu/gpuUtils.ts
Comment (cfchase) [inline backend/src/routes/api/gpu/gpuUtils.ts:106]: This await in a loop is going to be awfully slow.  Can we add all the promises to an array and do a Promises.all.  I mean it would even be faster to get all the machinesets with a limit and just filter out the ones that don't autoscale.

Comment (maroroman) [reply backend/src/routes/api/gpu/gpuUtils.ts]: Ok, added a loop that works with Promise.all

**Review by andrewballantyne — COMMENTED** [review]
Code looks okay to me. We may want to look at caching the value for a duration so the client can't spam this call, but that doesn't need to be in this PR. Not sure the best way to cache it -- FileWatcher might be okay, but a simple state variable in the file could work too... using timestamps and what not to determine if it should recompute the value. Custom logic is probably going to be annoying if it has issues.

**Comment by maroroman** [comment]
FYI: Added one more force push to deal with a log and fix the role yaml.

File: frontend/src/pages/notebookController/screens/server/GPUSelectField.tsx
Comment (andrewballantyne) [inline frontend/src/pages/notebookController/screens/server/GPUSelectField.tsx:31]:
```suggestion
        let availableScaleableGPU = 0;
        if (gpuInfo.autoscalers) {
          availableScaleableGPU = gpuInfo.autoscalers.reduce(
```

File: backend/src/routes/api/gpu/gpuUtils.ts
Comment (andrewballantyne) [inline backend/src/routes/api/gpu/gpuUtils.ts:52]:
```suggestion

  const data: GPUInfo = {
    configured: areGpusConfigured,
    available: maxGpuNumber,
    autoscalers: scalingLimit,
  };
  storage.lastFetch = Date.now();
  storage.lastValue = data;
  return data;
```

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

**Comment by andrewballantyne** [comment]
Added some data caching on the backend and removed the re-fetching toggle on the frontend.

**Comment by cfchase** [comment]
@maroroman we've made some progress here, but please test with a built image:
1. Can schedule a multi gpu notebook when a multi-gpu node is scaled down to 0 and there are no other gpu nodes but triggers a scaling event.
2. Can start a multi-gpu notebook when the available node only has 1 gpu available
3. Can start a single gpu notebook when 1 is available without scaling up

**Comment by cfchase** [comment]
Some comments on existing code:

There's another await in a for loop you need to Promise.all:
https://github.com/opendatahub-io/odh-dashboard/pull/573/files#diff-fe048868b84dce849bd698d4f237f445ecc70661f9ca1842d9a9429b4c2c37feR50

If that requests takes 3 seconds and there's 10 nodes, the user is going to be waiting for 30 seconds for the dialog to get a gpu value. It will grow O(n) as the cluster increases nodes.  I think this is affecting all current users.  Unless a request needs a previous result you shouldn't wait before executing. 

I think you also have to take out the `await` here: https://github.com/opendatahub-io/odh-dashboard/pull/573/files#diff-fe048868b84dce849bd698d4f237f445ecc70661f9ca1842d9a9429b4c2c37feR78

This token should be found at start up, added to the fastify.kube.saToken so you don't have to fetch it on every request
https://github.com/opendatahub-io/odh-dashboard/pull/573/files#diff-fe048868b84dce849bd698d4f237f445ecc70661f9ca1842d9a9429b4c2c37feR40

**Review by cfchase — APPROVED** [review]
/lgtm

**Review by andrewballantyne — APPROVED** [review]
/unhold

Forgot to unhold this after Chris approved it. Doing so now.

**Pull request merged**
